service/history/decision/task_handler_test.go

```diff
@@ -1131,6 +1134,200 @@ func TestHandleDecisionFailWorkflow(t *testing.T) {
+				taskHandler.mutableState.(*execution.MockMutableState).EXPECT().HasParentExecution().Return(true)
+				taskHandler.domainCache.(*cache.MockDomainCache).EXPECT().GetDomainName(executionInfo.ParentDomainID).Return("parent-domain-name", nil)
+				taskHandler.mutableState.(*execution.MockMutableState).EXPECT().AddContinueAsNewEvent(context.Background(), taskHandler.decisionTaskCompletedID, taskHandler.decisionTaskCompletedID, "parent-domain-name", attr).
+					Return(nil, nil, errors.New("some error adding continueAsNew event"))
+
+			},
+			asserts: func(t *testing.T, taskHandler *taskHandlerImpl, attr *types.ContinueAsNewWorkflowExecutionDecisionAttributes, err error) {
+				assert.NotNil(t, err)
+				assert.Equal(t, "some error adding continueAsNew event", err.Error())
+			},
+		},
+	}
+	for _, test := range tests {
+		t.Run(test.name, func(t *testing.T) {
+			taskHandler := newTaskHandlerForTest(t)
+			if test.expectMockCalls != nil {
+				test.expectMockCalls(taskHandler, test.attributes)
+			}
+			decision := &types.Decision{
+				DecisionType: func(i int32) *types.DecisionType {
+					decisionType := new(types.DecisionType)
+					*decisionType = types.DecisionType(i)
+					return decisionType
+				}(9), //types.DecisionTypeContinueAsNewWorkflowExecution
+				ContinueAsNewWorkflowExecutionDecisionAttributes: test.attributes,
+			}
+			err := taskHandler.handleDecision(context.Background(), decision)
+			test.asserts(t, taskHandler, test.attributes, err)
+		})
+	}
+}
+
+func TestValidateAttributes(t *testing.T) {
+	t.Run("failure", func(t *testing.T) {
+		taskHandler := newTaskHandlerForTest(t)
+		validationFn := func() error { return errors.New("some validation error") }
+		failedCause := new(types.DecisionTaskFailedCause)
+		err := taskHandler.validateDecisionAttr(validationFn, *failedCause)
+		assert.NotNil(t, err)
+		assert.Equal(t, "some validation error", err.Error())
+	})
+}
+
+func TestHandleDecisions(t *testing.T) {
+	t.Run("workflow size exceeds limit", func(t *testing.T) {
+		taskHandler := newTaskHandlerForTest(t)
+		taskHandler.sizeLimitChecker.historyCountLimitError = 10
+		taskHandler.mutableState.(*execution.MockMutableState).EXPECT().GetNextEventID().Return(int64(12)) //nextEventID - 1 > historyCountLimit of 10
+		taskHandler.mutableState.(*execution.MockMutableState).EXPECT().GetExecutionInfo().Return(&persistence.WorkflowExecutionInfo{})
+		taskHandler.mutableState.(*execution.MockMutableState).EXPECT().AddFailWorkflowEvent(taskHandler.sizeLimitChecker.completedID, &types.FailWorkflowExecutionDecisionAttributes{
+			Reason:  common.StringPtr(common.FailureReasonSizeExceedsLimit),
+			Details: []byte("Workflow history size / count exceeds limit."),
+		}).Return(nil, errors.New("some error adding fail workflow event"))
+		res, err := taskHandler.handleDecisions(context.Background(), []byte{}, []*types.Decision{})
+		assert.Nil(t, res)
+		assert.NotNil(t, err)
```

Comment:
nit: it's better to always check err first. I think, functions like `assert.ErrorContains()` will do both.

Reply:
Good to know 👍🏾